### PR TITLE
Probe to check if Spotbugs is enabled in a plugin

### LINKED ISSUE #151: Probe for SpotBugs
we should implement a probe to see if SpotBugs is configured/enabled

### PR DESCRIPTION
**Pull request by prince-panwar**
### Description

Resolves #151.

This Pull Request adds a new `SpotBugs` probe to the plugin codebase that checks whether a plugin has enabled spotbugs.

* SpotBugs is a static analysis tool that scans the plugin code for potential `bugs` and `security vulnerabilities`.
* we can ensure that the code is free from common bugs and vulnerabilities, reducing the risk of security breaches and other issues. 
* SpotBugs also provides detailed reports that can be used to identify and fix potential issues before they cause problems.

Overall, the use of SpotBugs in a plugin indicates that the plugin author is committed to producing a high quality product and is taking proactive steps to ensure the correctness and maintainability of the code. which can valuable information for the **plugin health scoring project**

### Submitter checklist

- [x] If the issue exists, it is well described and linked in the description
- [x] The description of this pull request is detailed and explains why this pull request is needed
- [ ] The changeset is on a specific branch
  - `feature/` for new feature, or improvements
  - `fix/` for bug fixes
  - `docs/` for any documentation changes
- [ ] If required, the documentation has been updated
- [ ] There is automated tests to cover the code change / addition
    - If there is no test, include an explanation why in the description
- [x] Run `mvn verify` locally and all tests are passing successfully
- [x] There is no new warnings (checkstyle nor spotbugs) on the code

### CODE REVIEW

**Comment by prince-panwar** [comment]
Since I am new to this project I might have made a few mistakes. Also, I have taken reference from the `ContributingGuide` plugin please review the code for any mistakes.
Looking forward i want to work on scoring for this probe and testing.

File: .devcontainer/DockerFile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM mcr.microsoft.com/devcontainers/java:0-17
```
Comment (alecharp) [inline .devcontainer/DockerFile:1]: This is not related to the described pull request. Please remove this file.

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
```
Comment (alecharp) [inline .devcontainer/devcontainer.json:1]: Same as above.

File: .devcontainer/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+version: '3.8'
```
Comment (alecharp) [inline .devcontainer/docker-compose.yml:1]: Same.

File: src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    protected ProbeResult doApply(Plugin plugin, ProbeContext context) {
+        if (plugin.getDetails().get(SCMLinkValidationProbe.KEY) == null) {
+            LOGGER.error("Couldn't run {} on {} because previous SCMLinkValidationProbe has null value in database", key(), plugin.getName());
+            return ProbeResult.error(key(), "SCM link has not been validated yet");
+        }
+    final Path repository = context.getScmRepository();
+    try (Stream<Path> paths = Files.find(repository, 2,
+                (file, basicFileAttributes) -> Files.isReadable(file)
+                        && ("pom.xml".equalsIgnoreCase(file.getFileName().toString())))) {
+            Optional<Path> pomFile = paths.findFirst();
```
Comment (alecharp) [inline src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java:36]: This would discriminate plugins using `Gradle` as build tool.

File: src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        if (pomFile.isPresent()) {
+            Path file = pomFile.get();
+              if (Files.lines(file).anyMatch(line -> line.contains("<groupId>com.github.spotbugs</groupId>"))) {
```
Comment (alecharp) [inline src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java:39]: Would it be possible that a plugin has this in a `pluginManagement` section and not in the `plugins` section, so in fact, spotbugs is not executed?

**Review by alecharp — CHANGES_REQUESTED** [review]
Thank you for trying to tackle this.

Please only include in a pull request files and modifications which are related to the pull request purpose. This helps understanding code addition and source code history.

Please note my comment on the probe implementation. I do believe that we would benefit from fetching the details from a GitHub Check Run rather than the `pom.xml` file.

**Comment by alecharp** [comment]
@prince-panwar have you been able to look at my comments? I think you could look at https://github.com/jenkins-infra/plugin-health-scoring/blob/main/core/src/main/java/io/jenkins/pluginhealth/scoring/probes/CodeCoverageProbe.java to see how to get the Check Runs from GitHub API.

**Comment by prince-panwar** [comment]
@alecharp thanks for the suggestion I will t do the changes and make a commit today and sorry for the late reply I got a little distracted because of a major festival here in India

File: src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java (hunk 4 above)
Comment (prince-panwar) [reply src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java:36]: I was under impression that the Jenkins plugin uses only maven as a build tool. Is it only the majority that uses maven?

File: src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java (hunk 5 above)
Comment (prince-panwar) [reply src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java:39]: Yeah, you are right need to change this too. thanks

Comment (prince-panwar) [reply src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java:39]: but I think if we are going to check from the GitHub check run then we would no longer need this code, do we?

**Comment by alecharp** [comment]
Can you please provide a test for this new probe? Thanks.

File: src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java (hunk 5 above)
Comment (alecharp) [reply src/main/java/io/jenkins/pluginhealth/scoring/probes/SpotBugsProbe.java]: correct.

**Review by alecharp — CHANGES_REQUESTED** [review]
I didn't catch that but your Probe needs to be in `core/src/main/java` not in `src/main/java`. Please move the file to the correct location.

**Comment by prince-panwar** [comment]
> I didn't catch that but your Probe needs to be in `core/src/main/java` not in `src/main/java`. Please move the file to the correct location.

I am confused here I have placed the probe where all the other plugins are.
I am not sure what you mean by `core/src/main/java`

**Comment by alecharp** [comment]
> I am confused here I have placed the probe where all the other plugins are.
I am not sure what you mean by core/src/main/java

Please see that all the other probes are now in `core` module. Please update your local repository.

**Comment by prince-panwar** [comment]
> I reformatted the source code, I didn't want to delay the merge of this PR just for that.

thank you . Now should I move ahead with scoring for this probe?

**Comment by alecharp** [comment]
I was thinking this probe could have a separate scoring implementation which would include code coverage, Jenkins core requirement and others like that in the future. I'll be creating a dedicated issue for that purpose.
